-datadir or -datadir="" option implies default datadir

hebasto opened the pull request:
This PR introduces a new behavior for `-datadir` (w/o a value) or `-datadir=""`: it means "use the default one". It allows to unset a `datadir` option specified in the config file by passing `-datadir` or `-datadir=""` as a command line option.

Credits:
This PR is inspired by **ryanofsky**'s [idea](https://github.com/bitcoin/bitcoin/pull/15864#discussion_r281785729):
> If somebody has a `datadir` option specified in the config file, but wants to unset it on the command line by passing `-nodatadir` or `-datadir=""`, it seems like this should be allowed.

The more general approach, however, has been described by **ryanofsky** [here](https://github.com/bitcoin/bitcoin/pull/16097#discussion_r288078111).

hebasto commented:
ping @ryanofsky

MarcoFalke commented on src/util/system.cpp:
```diff
@@ -741,7 +741,9 @@ const fs::path &GetDataDir(bool fNetSpecific)
     if (!path.empty()) return path;
 
-    if (gArgs.IsArgSet("-datadir")) {
+    tfm::format(std::cerr, "HEBASTO: -datadir negation is %s\n", gArgs.IsArgNegated("-datadir"));
```
:eyes:

hebasto replied:
I beg your pardon.

ryanofsky commented:
This seems ok. I think it could make sense for `-nodatadir` and `-datadir=` on the command line to override any datadir string in the configuration file and override any datadir in qt settings, and use default platform datadir as if no value was ever specified. 

Or alternately, it could make sense for -nodatadir to print an error, because it's similar enough to -nowallet, -nowalletdir, -nopid, -norpccookiefile options where you might want to be able to disable bitcoin from creating files & directories, but different from those options because bitcoin can't function without a datadir.

In either case, I think the PR description here should be updated to be more self contained. I think an ideal PR description would first say what the change in behavior is, then give reasons for the change, and only then go into deeper history and assigning credit for ideas, to be comprehensible as possible without digging into previous discussion.

hebasto commented:
@ryanofsky 

> In either case, I think the PR description here should be updated to be more self contained. I think an ideal PR description would first say what the change in behavior is, then give reasons for the change, and only then go into deeper history and assigning credit for ideas, to be comprehensible as possible without digging into previous discussion.

The PR description has been updated.

promag commented:
I would rather see just `-datadir` or `-datadir=""` as a way to set the default and override config.

`-nodatadir` doesn't make any sense to me.

promag commented:
Each arg could have a flag whether it can be negated or not. For instance `ArgsManager::AddArg(..., bool allow_negated = true)`.

hebasto commented:
@promag 
> I would rather see just `-datadir` or `-datadir=""` as a way to set the default and override config.
> `-nodatadir` doesn't make any sense to me.

You and @ryanofsky have convinced me. Going to implement your suggestion tonight.

> Each arg could have a flag whether it can be negated or not. For instance `ArgsManager::AddArg(..., bool allow_negated = true)`.

Could you look into #16097?

hebasto committed:
Add a way to set default -datadir

-datadir or -datadir="" in command line will set the default datadir 
(overriding config file options).

hebasto commented:
@ryanofsky and @promag Thank you for your reviews.
> I would rather see just `-datadir` or `-datadir=""` as a way to set the default and override config.

Done.

> ... it could make sense for `-nodatadir` to print an error...

I'd rather use a more general solution like #16097.

hebasto commented:
Closed in favor of #15864.

See: https://github.com/bitcoin/bitcoin/pull/15864/commits/740d41ce9f7fdf209366e930bd0fdcc6b1bc6b79#r306906349
